---
title: "Azimuth annotation for COVID PBMCs data"
subtitle: 'Analysis for ASI/OSC Hackathon challenge #1'
author: "Angli XUE"
output:
  html_document:
    theme: united
    df_print: kable
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---
***
  
```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

# Overview 

This vignette demonstrates how to map scRNA-seq data from Liu et al. (Cell. 2021) to Azimuth reference datasets in R.

The code pipeline was forked from https://github.com/satijalab/azimuth/blob/HEAD/vignettes/run_azimuth_tutorial.Rmd

Reference-based mapping provides an attractive alternative to unsupervised analysis. When well-curated and annotated references are available, reference-mapping can rapdidly, robustly, and sensitively annotate and interpret query datasets. As part of the [Human Biomolecular Atlas Project](portal.hubmapconsortium.org), The Satija Team have built integrated references for multiple human tissues, available at [azimuth.hubmapconsortium.org](azimuth.hubmapconsortium.org). Azimuth is a web-tool that maps user-uploaded datasets - starting from an unnormalized expression counts matrix.

The Satija Team have recently released updated software that enables the Azimuth workflow to be replicated directly in R, without the need to upload data into the web application. Reference datasets are automatically downloaded as part of out previously released [SeuratData](https://github.com/satijalab/seurat-data) framework. Datasets can then be passed through Azimuth using a single command. 

As with the web application, Azimuth is compatible with a wide range of inputs, including Seurat objects, 10x HDF5 files, and Scanpy/h5ad files. Once Azimuth is run, a Seurat object is returned which contains

* Cell annotations (at multiple levels of resolution)
* Prediction scores (i.e. confidence scores) for each annotation
* Projection onto the reference-derived 2-dimensional UMAP visualization

In this vignette, we demonstrate the use of a function `RunAzimuth()` which facilitates annotation of single cell datasets.

<details>
  <summary>**View data download code**</summary>
  
  To download all the required files, please check the `/data` folder:
  
```{sh, eval=FALSE}
# Example code for illustration purposes only
wget https://figshare.com/ndownloader/files/27686835 -O human_cd34_bone_marrow.h5ad
```

</details>
  
# Installation
  
We first install and load Seurat, Azimuth, and Seurat-Data

```{r installlibraries, eval = FALSE}
devtools::install_github("satijalab/seurat-data")
devtools::install_github("satijalab/azimuth")
install.packages("Seurat")
BiocManager::install("glmGamPoi")
```

```{r loadlibraries}
library(Seurat)
library(Azimuth)
library(SeuratData)
library(patchwork)
library(glmGamPoi)
library(dplyr)
library(ggplot2)

library(knitr)
library(kableExtra)
library(rmarkdown)
source("../src/recode.R")
```

# Map COVID PBMC datasets based on scRNA-seq and/or CITE-seq

We consider a COVID dataset from [Time-resolved systems immunology reveals a late juncture linked to fatal COVID-19, Cell 2021](https://pubmed.ncbi.nlm.nih.gov/33713619/), which profiled a total of 372,081 human peripheral blood mononuclear cells (PBMCs) from 46 donors (33 COVID patients) using scRNA-seq and CITE-seq. The download links for the raw data are [here](https://docs.google.com/spreadsheets/d/1ktYSwelU0LKWua0OGWDQUt3GEHb2VtGHCW8ZvoyF5IE/edit?usp=sharing). We read the data in, and map it to the Azimuth [PBMC reference](https://azimuth.hubmapconsortium.org/references/#Human%20-%20PBMC).

Let's first visualize the data to replicate the Figure 2B in the paper.

```{r pbmc_dimplot_check, fig.width = 12, fig.height = 5}
## Read in Liu_et_al_Cell_2021_COVID data
rna1 = readRDS("../../data/Liu_et_al_Cell_2021_COVID/10_1016_j_cell_2021_02_018_innate_rna.rds")
rna2 = readRDS("../../data/Liu_et_al_Cell_2021_COVID/10_1016_j_cell_2021_02_018_adaptive_rna.rds")

prot1 = readRDS("../../data/Liu_et_al_Cell_2021_COVID/10_1016_j_cell_2021_02_018_innate_protein.rds")
prot2 = readRDS("../../data/Liu_et_al_Cell_2021_COVID/10_1016_j_cell_2021_02_018_adaptive_protein.rds")

# Rename the cell type
prot1$cell_type_abbr = rename_cells(prot1$cell_type)
prot2$cell_type_abbr = rename_cells(prot2$cell_type)

p1 <- DimPlot(prot1, group.by = "cell_type_abbr", label = TRUE, label.size = 3) + NoLegend() + ggtitle("Innate cell only: 125,117 cells")
p2 <- DimPlot(prot2, group.by = "cell_type_abbr", label = TRUE, label.size = 3) + NoLegend() + ggtitle("Adaptive cell only: 246,964 cells")
p1 + p2
```

The UMAP plots are the same as Figure 2B. Then we start to explore the datasets in more details

Here are the cell type names and corresponding ontology IDs
```{r check cell type ID, echo = FALSE}
ct = data.frame(cell_type = c(prot1$cell_type,prot2$cell_type),cell_type_ontology_term_id = c(prot1$cell_type_ontology_term_id,prot2$cell_type_ontology_term_id))
ct = ct[!duplicated(ct$cell_type),]
rownames(ct) = 1:nrow(ct)
ct$cell_type_abbr = rename_cells(ct$cell_type)
ct = ct[,c(1,3,2)]

ct %>%
  kbl() %>%
  kable_styling(bootstrap_options = "striped", full_width = F, position = "left")
```


```{r data_explore}
# Note that since measurements were made in the same cells, the two matrices have identical column names
all.equal(colnames(rna1), colnames(prot1))
all.equal(colnames(rna2), colnames(prot2))

# The counts slot for rna data is empty. Why? Save space?
dim(rna1@assays$RNA@counts)
# Assign the data slot back to counts slot
rna1@assays$RNA@counts = rna1@assays$RNA@data
rna2@assays$RNA@counts = rna2@assays$RNA@data

# I remember the data slot should store the log-normalized counts
# Don't know why these two slots are equivalent
# https://github.com/satijalab/seurat/issues/3711
all(prot1@assays$ADT@data == prot1@assays$ADT@counts)
all(prot2@assays$ADT@data == prot2@assays$ADT@counts)

# Set a key for the assays otherwise will get error in vapply
# https://github.com/satijalab/seurat/issues/5676
rna1@assays$RNA@key = "rna1_"
prot1@assays$ADT@key = "adt1_"
rna2@assays$RNA@key = "rna2_"
prot2@assays$ADT@key = "adt2_"

# Merge the RNA and protein data separately first
data1 <- merge(rna1, y = rna2, add.cell.ids = c("Innate", "Adaptive"))
data2 <- merge(prot1, y = prot2, add.cell.ids = c("Innate", "Adaptive"))
```

Clustering the data based on RNA-seq

```{r RNA_assay, results='hide'}
data1 <- NormalizeData(data1)
data1 <- FindVariableFeatures(data1)
data1 <- ScaleData(data1)
data1 <- RunPCA(data1, verbose = TRUE)
data1 <- FindNeighbors(data1, dims = 1:15)
data1 <- FindClusters(data1, resolution = 0.5, verbose = TRUE)
data1 <- RunUMAP(data1, dims = 1:30)
```

Run the main function of Azimuth annotation based on RNA-seq

```{r run_Azimuth_by_RNA-seq, results='hide'}
# The RunAzimuth function can take a Seurat object as input
DefaultAssay(data1) <- "RNA"
data1 <- RunAzimuth(data1, assay = "RNA", reference = "pbmcref")
saveRDS(data1,"../../results/innate_adapative_rna_processed.rds")
```

We can now visualize the outputs of Azimuth. Previously this dataset was annotated with 10 cell types for innate cells and 14 cell types for adaptive cells.

```{r pbmc_dimplot_RNA, fig.width=12, fig.height=5}
# data1 <- readRDS("../results/innate_adapative_rna_processed.rds")
p1 <- DimPlot(data1, group.by = "cell_type", label = TRUE, label.size = 3) + NoLegend() + ggtitle("Previous annotation")
p2 <- DimPlot(data1, group.by = "predicted.celltype.l2", label = TRUE, label.size = 3) + NoLegend() + ggtitle("Current RNA annotation")
p1 + p2
```

Question: can we use the ADT assay to annotate the cell type?

```{r ADT_assay, results='hide'}
data2 <- NormalizeData(data2, normalization.method = 'CLR', margin = 2)
data2 <- FindVariableFeatures(data2)
data2 <- ScaleData(data2)
data2 <- RunPCA(data2, verbose = TRUE)
data2 <- FindNeighbors(data2, dims = 1:15)
data2 <- FindClusters(data2, resolution = 0.5, verbose = TRUE)
data2 <- RunUMAP(data2, dims = 1:30)
```

Run the main function of Azimuth annotation based on CITE-seq

```{r run_Azimuth_by_CITE-seq, results='hide'}
# The RunAzimuth function can take a Seurat object as input
DefaultAssay(data2) <- "ADT"
data2 <- RunAzimuth(data2, assay = "ADT", reference = "pbmcref")
saveRDS(data2,"../../results/innate_adaptive_protein_processed.rds")
```

We can now visualize the outputs of Azimuth. Previously this dataset was annotated with 10 cell types for innate cells and 14 cell types for adaptive cells.

```{r pbmc_dimplot_ADT, fig.width=12, fig.height=5}
p1 <- DimPlot(data2, group.by = "cell_type", label = TRUE, label.size = 3) + NoLegend() + ggtitle("Previous annotation")
p2 <- DimPlot(data2, group.by = "predicted.celltype.l2", label = TRUE, label.size = 3) + NoLegend() + ggtitle("New ADT annotation")
p1 + p2
```

# Compare the annotation based on RNA only and protein only

Since the clustering parameters are not 100% same as in the literature, we just compare the annotation using RNA or protein in the current workflow.

```{r pbmc_dimplot_compare, fig.width=12, fig.height=5}
p1 <- DimPlot(data1, group.by = "predicted.celltype.l2", label = TRUE, label.size = 3) + NoLegend() + ggtitle("RNA annotation")
p2 <- DimPlot(data2, group.by = "predicted.celltype.l2", label = TRUE, label.size = 3) + NoLegend() + ggtitle("Protein annotation")
p1 + p2
```

We can visualize the expression of canonical marker genes to examine the accuracy of predictions. Note that Azimuth normalizes data (internally) before mapping, but does not return the results, so we normalize the data here before visualization.

Here, we specifically visualize:

* The expression of CCR7 on CD4 and CD8 Naive T cells
* The expression of FCGR3A on CD16+ monocytes, CD56dim NK cells, and cytotoxic CD8 T cells
* The expression of AXL on rare populations of AXL+SIGLEC6+ dendritic cells ([ASDC](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5775029/))
* Prediction scores for the annotation CD4+ regulatory T cells (Treg)

```{r pbmc_res1, fig.width = 12, fig.height = 10}
data1 <- NormalizeData(data1)
Idents(data1) <- "predicted.celltype.l2"
p1 <- FeaturePlot(data1, features = "CCR7")
p2 <- FeaturePlot(data1, features = "FCGR3A")
p3 <- VlnPlot(data1, features = "AXL", group.by = "predicted.celltype.l2", idents = c("ASDC", "pDC", "cDC1", "cDC2"))
p4 <- FeaturePlot(data1, features = "predictionscorecelltypel2_Treg")
p1 + p2 + p3 + p4 + plot_layout(ncol=2)
```

<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>