---
title: "Azimuth annotation for COVID PBMCs data"
subtitle: 'Analysis for ASI/OSC Hackathon challenge #1'
author: "Angli XUE"
output:
  html_document:
    theme: united
    df_print: kable
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---
***
  
```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

# Overview 

This vignette demonstrates how to map scRNA-seq data from Liu et al. (Cell. 2021) to Azimuth reference datasets in R.

The code pipeline was forked from https://github.com/satijalab/azimuth/blob/HEAD/vignettes/run_azimuth_tutorial.Rmd

Reference-based mapping provides an attractive alternative to unsupervised analysis. When well-curated and annotated references are available, reference-mapping can rapdidly, robustly, and sensitively annotate and interpret query datasets. As part of the [Human Biomolecular Atlas Project](portal.hubmapconsortium.org), The Satija Team have built integrated references for multiple human tissues, available at [azimuth.hubmapconsortium.org](azimuth.hubmapconsortium.org). Azimuth is a web-tool that maps user-uploaded datasets - starting from an unnormalized expression counts matrix.

The Satija Team have recently released updated software that enables the Azimuth workflow to be replicated directly in R, without the need to upload data into the web application. Reference datasets are automatically downloaded as part of out previously released [SeuratData](https://github.com/satijalab/seurat-data) framework. Datasets can then be passed through Azimuth using a single command. 

As with the web application, Azimuth is compatible with a wide range of inputs, including Seurat objects, 10x HDF5 files, and Scanpy/h5ad files. Once Azimuth is run, a Seurat object is returned which contains

* Cell annotations (at multiple levels of resolution)
* Prediction scores (i.e. confidence scores) for each annotation
* Projection onto the reference-derived 2-dimensional UMAP visualization

In this vignette, we demonstrate the use of a function `RunAzimuth()` which facilitates annotation of single cell datasets.

<details>
  <summary>**View data download code**</summary>
  
  To download all the required files, you can run the following lines in a shell:
  
```{sh, eval=FALSE}
wget https://figshare.com/ndownloader/files/27686835 -O human_cd34_bone_marrow.h5ad
```

</details>
  
# Installation
  
We first install and load Seurat, Azimuth, and Seurat-Data

```{r installlibraries, eval=FALSE}
devtools::install_github("satijalab/seurat-data")
devtools::install_github("satijalab/azimuth")
install.packages("Seurat")
BiocManager::install("glmGamPoi")
```

```{r loadlibraries}
library(Seurat)
library(Azimuth)
library(SeuratData)
library(patchwork)
library(glmGamPoi)
library(dplyr)
library(ggplot2)
source("./src/recode.R")
```

## Map COVID PBMC datasets based on scRNA-seq and/or CITE-seq

We consider a COVID dataset from [Time-resolved systems immunology reveals a late juncture linked to fatal COVID-19, Cell 2021](https://pubmed.ncbi.nlm.nih.gov/33713619/), which profiled a total of 372,081 human peripheral blood mononuclear cells (PBMCs) from 46 donors (33 COVID patients) using scRNA-seq and CITE-seq. The download links for the raw data are [here](https://docs.google.com/spreadsheets/d/1ktYSwelU0LKWua0OGWDQUt3GEHb2VtGHCW8ZvoyF5IE/edit?usp=sharing). We read the data in, and map it to the Azimuth [PBMC reference](https://azimuth.hubmapconsortium.org/references/#Human%20-%20PBMC).

Let's first visualize the data to replicate the Figure 2B in the paper.

```{r pbmc_dimplot, fig.width=12, fig.height=5}
## Read in Liu_et_al_Cell_2021_COVID data
rna1 = readRDS("../data/Liu_et_al_Cell_2021_COVID/10_1016_j_cell_2021_02_018_innate_rna.rds")
rna2 = readRDS("../data/Liu_et_al_Cell_2021_COVID/10_1016_j_cell_2021_02_018_adaptive_rna.rds")

prot1 = readRDS("../data/Liu_et_al_Cell_2021_COVID/10_1016_j_cell_2021_02_018_innate_protein.rds")
prot2 = readRDS("../data/Liu_et_al_Cell_2021_COVID/10_1016_j_cell_2021_02_018_adaptive_protein.rds")

p1 <- DimPlot(rna1, group.by = "cell_type", label = TRUE, label.size = 3) + NoLegend() + ggtitle("Innate cell only: 125,117 cells")
p2 <- DimPlot(rna2, group.by = "cell_type", label = TRUE, label.size = 3) + NoLegend() + ggtitle("Adaptive cell only: 246,964 cells")
p1 + p2
```

The UMAP plots are the same as Figure 2B. Then we start to explore the datasets in more details

```{r data_explore, results='hide'}
# Note that since measurements were made in the same cells, the two matrices have identical column names
all.equal(colnames(rna1), colnames(prot1))
all.equal(colnames(rna2), colnames(prot2))

# The counts slot for rna data is empty. Why? Save space?
dim(rna1@assays$RNA@counts)
# Assign the data slot back to counts slot
rna1@assays$RNA@counts = rna1@assays$RNA@data
rna2@assays$RNA@counts = rna2@assays$RNA@data

# I remember the data slot should store the log-normalized counts
# Don't know why these two slots are equivalent
# https://github.com/satijalab/seurat/issues/3711
all(prot1@assays$ADT@data == prot1@assays$ADT@counts)
all(prot2@assays$ADT@data == prot2@assays$ADT@counts)

# Set a key for the assays otherwise will get error in vapply
# https://github.com/satijalab/seurat/issues/5676
rna1@assays$RNA@key = "rna1_"
prot1@assays$ADT@key = "adt1_"
rna2@assays$RNA@key = "rna2_"
prot2@assays$ADT@key = "adt2_"
```

Question: can we use the ADT assay to annotate the cell type?

```{r ADT_assay, results='hide'}

# Question: can we use the ADT assay to annotate the cell type?

prot1 <- NormalizeData(prot1, normalization.method = 'CLR', margin = 2)
prot1 <- FindVariableFeatures(prot1)
prot1 <- ScaleData(prot1)
prot1 <- RunPCA(prot1, verbose = TRUE)
prot1 <- FindNeighbors(prot1, dims = 1:15)
prot1 <- FindClusters(prot1, resolution = 1, verbose = TRUE)
prot1 <- RunUMAP(prot1, dims = 1:30)

# Add the ADT assay to the RNA Seurat object
# The ADT assay stands for "antibody-derived tags"
# adt_assay <- CreateAssayObject(counts = prot1@assays$ADT@data)
# rna1[["ADT"]] <- adt_assay

```


Run the main function of Azimuth annotation

```{r run Azimuth, results='hide'}
# The RunAzimuth function can take a Seurat object as input
DefaultAssay(prot1) <- "ADT"
prot1 <- RunAzimuth(prot1, assay = "ADT", reference = "pbmcref")
```

We can now visualize the outputs of Azimuth. Note that all cells are labeled with high-resolution annotations. Previously this dataset was annotated with 10 cell types for innate cells and 14 cell types for adaptive cells.

```{r pbmc_dimplot, fig.width=12, fig.height=5}

# data2$cell_type <- rename_cells2(data2$cell_type)

p1 <- DimPlot(data1, group.by = "cell_type", label = TRUE, label.size = 3) + NoLegend() + ggtitle("Innate cell only: 125,117 cells")
p2 <- DimPlot(data1, group.by = "ethnicity")
p1 + p2
```

We can visualize the expression of canonical marker genes to examine the accuracy of predictions. Note that Azimuth normalizes data (internally) before mapping, but does not return the results, so we normalize the data here before visualization.

Here, we specifically visualize:

* The expression of CCR7 on CD4 and CD8 Naive T cells
* The expression of FCGR3A on CD16+ monocytes, CD56dim NK cells, and cytotoxic CD8 T cells
* The expression of AXL on rare populations of AXL+SIGLEC6+ dendritic cells ([ASDC](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5775029/))
* Prediction scores for the annotation CD4+ regulatory T cells (Treg)

```{r pbmc_res1, fig.width=12, fig.height=10}
pbmcsca <- NormalizeData(pbmcsca)
Idents(pbmcsca) <- "predicted.celltype.l2"
p1 <- FeaturePlot(pbmcsca, features = "CCR7")
p2 <- FeaturePlot(pbmcsca, features = "FCGR3A")
p3 <- VlnPlot(pbmcsca, features = "AXL", group.by = "predicted.celltype.l2", idents = c("ASDC", "pDC", "cDC1", "cDC2"))
p4 <- FeaturePlot(pbmcsca, features = "predictionscorecelltypel2_Treg")
p1 + p2 + p3 + p4 + plot_layout(ncol=2)
```

## See available references

You can search all available datasets in SeuratData (focusing on Azimuth references)

```{r show_avail}
available_data <- AvailableData()
available_data[grep("Azimuth", available_data[,3]), 1:3]
```

## Map CD34+ cells from Human Bone Marrow (stored as an h5ad file)

As a final example, we map data from CD34+ human bone marrow cells from the manuscript [Characterization of cell fate probabilities in single-cell data with Palantir
](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7549125/). We map to our [Azimuth Human Bone Marrow reference](https://azimuth.hubmapconsortium.org/references/#Human%20-%20Bone%20Marrow), which includes both progenitor and differentiated cells.

Azimuth can also take the path to an h5ad object as input. In this case, Azimuth extracts the unnormalized counts from the object, and proceeds with mapping.

```{r bm_annotate, message=FALSE, warning=FALSE, results='hide'}
# The default timeout is too short (60s) and the download will time out
options(timeout = 1000)

bm <- RunAzimuth(
  query = "../data/human_cd34_bone_marrow.h5ad",
  reference = "bonemarrowref"
)
```

As expected, query cells map to CD34+ celltypes which represent a subset of celltypes present in the reference. Rare cells that map to differentiated cell populations (i.e. CD4 memory), map with low prediction scores. Reference-mapping also harmonizes two separate runs.

```{r bm_plots, fig.width=10, fig.height=10}
p1 <- DimPlot(bm, group.by = "predicted.celltype.l2", label = TRUE) + NoLegend()
p2 <- DimPlot(bm, group.by = "orig.ident") 
sort(table(bm$predicted.celltype.l2), decreasing = TRUE)
p3 <- VlnPlot(bm, features = "predicted.celltype.l2.score") + NoLegend()
(p1 + p2) / p3
```

Lastly, we can visualize markers associated with lineage differentiation to verify our annotations including AVP (HSC), KLF1 (erythroid), MPO (myeloid), and VWF (platelet).

```{r bm_feature_plots, fig.width=12, fig.height=10}
# normalize before visualization
bm <- NormalizeData(bm)
FeaturePlot(bm, features = c("AVP", "KLF1", "MPO", "VWF"))
```

<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>